public/service/metrics.go

```diff
@@ -125,12 +135,24 @@ type MetricsExporterGaugeCtor func(labelValues ...string) MetricsExporterGauge
 // MetricsExporterCounter represents a counter metric of a given name and
 // labels.
 type MetricsExporterCounter interface {
```

Comment:
Modifying this interface is another breaking change, it means any existing implementations of `MetricsExporterCounter` will fail to compile if we merge this PR, which we can't have.

The backwards compatible way to do it would be to check for `interface{ IncrFloat64(count float64) }` and so on inside the airgap and fallback to the integer equivalents when it isn't implemented.

Reply:
@Jeffail I'm sorry but I didn't fully get that, why would it fail to compile since current and previous versions of Benthos simply lack these interfaces and these newly added interfaces are implemented where needed(airGapGauge, airGapCounter...)?

I created a local sample application that makes use of these changes, that's a testing config file
```
http:
  enabled: true
  address: 0.0.0.0:4195
  root_path: /
  debug_endpoints: false

input:
  label: ""
  generate:
    mapping: |
      root.msg = "hello world"
      root.value = 10.2
    interval: 1s

pipeline:
  threads: -1
  processors:
    - metric:
        type: "counter_by"
        name: "hello_metric"
        value: ${! this.value }

output:
  drop: { }

metrics:
  prometheus:
    use_histogram_timing: true
    add_process_metrics: true
    add_go_metrics: true

shutdown_timeout: 20s
```
and the following is an extract of `go.mod` to make use of my changes:
```
module benthos-metrics

go 1.21

replace github.com/benthosdev/benthos/v4 v4.21.0 => /Users/fouad/workspace/personal/benthos

require github.com/benthosdev/benthos/v4 v4.21.0

require (
	...
)
```

even though Benthos version `4.21.0` lacks all of these changes, the application compiles and runs as expected.

@Jeffail I'd really appreciate it if you can elaborate a little bit more on this

Reply:
@fouadkada if a plugin author has implemented the old `MetricsExporterCounter` and then you add more methods to it their implementation will no longer be valid. Anyone using that hypothetical plugin will have a failed build once they upgrade to the version of benthos containing your expanded version.

This stuff may seem minor, as it's only a quick fix for the plugin author to implement the new methods. However, the reality is these changes can be extremely problematic as it breaks semantic versioning and a module that imports both benthos among a few plugins doesn't have control over whether/when the plugin authors can update to your new interface.

There is no backwards compatible way to expand or change a public interface. You need to instead allow plugin authors to optionally implement the expanded set of methods and then internally make those checks yourself (or fall back in cases where it's not implemented).

If you're not sure how to do that then that's fine, I can do it myself when I next get a chance.

Reply:
@Jeffail that makes a lot more sense now. The moment you mentioned custom plugin implementing the `MetricsExporterCounter` interface then I can now see why it's a breaking change.

I'd really appreciate it you can expand a little bit more on the following with an example or something then I can try to implement it myself:
> The backwards compatible way to do it would be to check for interface{ IncrFloat64(count float64) } and so on > inside the airgap and fallback to the integer equivalents when it isn't implemented.

Reply:
Okay this is next on my priority list for PRs, I'll work through the breaking changes myself as I get it merged as it's hard to describe without just writing it.